### PR TITLE
Remove extra spaces from JSON output.

### PR DESCRIPTION
**Pull request by psclafer**
The JSON us mainly used for storage or network communications. Since micropython aims to be as unbloated as possible to be used for small applications like IoT, the JSON output should be as compact as possible.

### CODE REVIEW

**Comment by roadsidedan** [comment]
How  do  I stop this chain email shitSent from my T-Mobile 4G LTE Device
-------- Original message --------From: Pascal S <notifications@github.com> Date: 11/5/18  9:25 AM  (GMT-05:00) To: micropython/micropython <micropython@noreply.github.com> Cc: Subscribed <subscribed@noreply.github.com> Subject: [micropython/micropython] Remove extra spaces from JSON output.
  (#4294) The JSON us mainly used for storage or network communications. Since micropython aims to be as unbloated as possible to be used for small applications like IoT, the JSON output should be as compact as possible.



You can view, comment on, or merge this pull request online at:
  https://github.com/micropython/micropython/pull/4294

Commit Summary

  Remove extra spaces from JSON output.


File Changes

  
    M
    py/objdict.c
    (8)
  
  
    M
    py/objlist.c
    (5)
  
  
    M
    py/objtuple.c
    (5)
  


Patch Links:

  https://github.com/micropython/micropython/pull/4294.patch
  https://github.com/micropython/micropython/pull/4294.diff


—You are receiving this because you are subscribed to this thread.Reply to this email directly, view it on GitHub, or mute the thread.
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/micropython/micropython","title":"micropython/micropython","subtitle":"GitHub repository","main_image_url":"https://assets-cdn.github.com/images/email/message_cards/header.png","avatar_image_url":"https://assets-cdn.github.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/micropython/micropython"}},"updates":{"snippets":[{"icon":"DESCRIPTION","message":"Remove extra spaces from JSON output. (#4294)"}],"action":{"name":"View Pull Request","url":"https://github.com/micropython/micropython/pull/4294"}}}
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/micropython/micropython/pull/4294",
"url": "https://github.com/micropython/micropython/pull/4294",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
},
{
"@type": "MessageCard",
"@context": "http://schema.org/extensions",
"hideOriginalBody": "false",
"originator": "AF6C5A86-E920-430C-9C59-A73278B5EFEB",
"title": "Remove extra spaces from JSON output. (#4294)",
"sections": [
{
"text": "",
"activityTitle": "**Pascal S**",
"activityImage": "https://assets-cdn.github.com/images/email/message_cards/avatar.png",
"activitySubtitle": "@psclafer",
"facts": [

]
},
{
"title": "Commit Summary",
"facts": [
{
"name": "195970d",
"value": "Remove extra spaces from JSON output."
}
]
},
{
"title": "File Changes",
"facts": [
{
"name": "Modified",
"value": "[py/objdict.c](https://github.com/micropython/micropython/pull/4294/files#diff-0) (8 changes)"
},
{
"name": "Modified",
"value": "[py/objlist.c](https://github.com/micropython/micropython/pull/4294/files#diff-1) (5 changes)"
},
{
"name": "Modified",
"value": "[py/objtuple.c](https://github.com/micropython/micropython/pull/4294/files#diff-2) (5 changes)"
}
]
}
],
"potentialAction": [
{
"name": "Add a comment",
"@type": "ActionCard",
"inputs": [
{
"isMultiLine": true,
"@type": "TextInput",
"id": "IssueComment",
"isRequired": false
}
],
"actions": [
{
"name": "Comment",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"IssueComment\",\n\"repositoryFullName\": \"micropython/micropython\",\n\"issueId\": 4294,\n\"IssueComment\": \"{{IssueComment.value}}\"\n}"
}
]
},
{
"name": "Close pull request",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"PullRequestClose\",\n\"repositoryFullName\": \"micropython/micropython\",\n\"pullRequestId\": 4294\n}"
},
{
"targets": [
{
"os": "default",
"uri": "https://github.com/micropython/micropython/pull/4294"
}
],
"@type": "OpenUri",
"name": "View on GitHub"
},
{
"targets": [
{
"os": "default",
"uri": "https://github.com/micropython/micropython/pull/4294.patch"
}
],
"@type": "OpenUri",
"name": "View patch"
},
{
"targets": [
{
"os": "default",
"uri": "https://github.com/micropython/micropython/pull/4294.diff"
}
],
"@type": "OpenUri",
"name": "View diff"
},
{
"name": "Unsubscribe",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"MuteNotification\",\n\"threadId\": 406465587\n}"
}
],
"themeColor": "26292E"
}
]

**Comment by peterhinch** [comment]
Click on your GitHub avatar, settings - notifications - uncheck the email options.

**Review by pfalcon — CHANGES_REQUESTED** [review]
> The JSON us mainly used for storage or network communications. 

How comes? The purpose of JSON is to be *human-readable* *interchange* format. Nobody uses it when size efficiency is needed, and uses binary-based formats instead.

> Since micropython aims to be as unbloated as possible to be used for small applications like IoT, the JSON output should be as compact as possible.

MicroPython's aim is also what can be called "understandable programming". Again, JSON is human-readable format, and purpose of this patch can be seen as an obfuscation measure. It's not good for when someone develops/debug their own applications. It's not good when someone reviews what someone else's module is doing on their system.

On a wider perspective, some people want the most compact JSON, some want the most readable "pretty-print" format (one value per line, line breaks, indentation), some pretty-print, but with custom rules (like, particular object's fields encoded on one line; appearing in particular order), etc., etc. MicroPython's default JSON formatter can't satisfy all those needs, as MicroPython is indeed optimized for size. In this regard, MicroPython's current format is pretty good compromise.

Custom modules can be developed to other needs (e.g., micropython-lib includes CPython's "json" module implementation which offers more option (still not covering all possible usecases)).

Summing up, I'm -0.8 on this change.

**Comment by dpgeorge** [comment]
Thanks for the contribution.  In this case the change diverges from CPython (the default is to have spaces in the JSON output) so that's a reason not to merge it.

If this feature is needed then it would be better to implement proper options to JSON printing for the user to change the separator, as per CPython.
